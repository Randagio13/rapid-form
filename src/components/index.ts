export { default as Button } from './button/Button'
export { default as Form } from './form/Form'
export { default as Text } from './input/Text'
