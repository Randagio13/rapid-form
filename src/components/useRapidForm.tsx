import { useCallback, useDebugValue, useReducer } from 'react'

const setErrors = (data: any, dispatch: any) => {
  if (!data.value) {
    dispatch({
      type: 'empty',
      data: {
        [data.name]: {
          ...data
        }
      },
      errors: {
        [data.name]: {
          message: `${data.name} is required`
        }
      }
    })
  } else {
    dispatch({
      type: 'change',
      data: {
        [data.name]: {
          ...data
        }
      }
    })
  }
}

const fetchReducer = (state: any, action: any) => {
  console.log('state fetch :', state)
  if (action.type === 'change') {
    Object.keys(state.errors).map((k: any) => {
      if (state.data.hasOwnProperty(k)) {
        delete state.errors[k]
      }
    })
    return {
      ...state,
      data: {
        ...state.data,
        ...action.data
      },
      errors: {
        ...state.errors
      }
    }
  } else if (action.type === 'empty') {
    return {
      ...state,
      data: {
        ...state.data,
        ...action.data
      },
      errors: {
        ...state.errors,
        ...action.errors
      }
    }
  } else {
    throw new Error(`That action type isn't supported.`)
  }
}

export const useValidation = (dispatch: any) =>
  useCallback(
    (ref: any) => {
      ref.oninput = (e: React.ChangeEvent<HTMLInputElement>) => {
        const v = e.target.value
        const n = e.target.name
        const t = e.target.type
        const r = e.target.required
        const data: any = {
          [n]: {
            value: v,
            name: n,
            type: t,
            required: r
          }
        }
        setErrors(data[n], dispatch)
      }
    },
    [dispatch]
  )

export default function useRapidForm(): any {
  const [state, dispatch] = useReducer(fetchReducer, {
    data: {},
    errors: {}
  })
  const validation = useValidation(dispatch)
  useDebugValue(state.data)
  return {
    data: state.data,
    errors: state.errors,
    validation
  }
}
