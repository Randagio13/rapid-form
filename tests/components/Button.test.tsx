// TODO: Added mock Redux component
import { Button } from 'components'
import { shallow } from 'enzyme'
import * as React from 'react'
import { create } from 'react-test-renderer'

const component = create(
  <Button type='submit'>{'Sign up'}</Button>
).toJSON()

test('Snapshot', () => {
  // expect(component).toMatchSnapshot()
})

test('Check property component', () => {
  expect(component.type).toEqual('button')
  expect(component.props.type).toEqual('submit')
  expect(component.children[0]).toEqual('Sign up')
})
