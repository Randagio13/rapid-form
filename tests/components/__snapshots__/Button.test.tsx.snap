// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
<button
  type="submit"
>
  Sign up
</button>
`;
